### PR TITLE
Enabled bfloat16 for cuda

### PR DESCRIPTION
**Pull request by izdeby**
Enabled basic support for bfloat16 on cuda
Tested via unit tests

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
This diff looks very familiar to me. Is it a respin of a diff you did previously?

File: aten/src/THC/THCNumerics.cuh [hunk 1]
```diff
@@ -309,6 +309,77 @@ struct THCNumerics<float> {
...
+    #ifdef _MSC_VER
+      // Windows requires this explicit conversion. The reason is unclear
+      // related issue with clang: https://reviews.llvm.org/D37906
+      return ::isinf((float) a);
+    #else
+      return ::isinf(a);
+    #endif
+  }
+};
+
```
Comment (izdeby) [inline aten/src/THC/THCNumerics.cuh]: @ezyang, this part

Comment (izdeby) [reply aten/src/THC/THCNumerics.cuh]: `This diff looks very familiar to me. Is it a respin of a diff you did previously?`
The diff is identical to the one that have been reverted except this part which fixes win build.

File: aten/src/THC/THCNumerics.cuh [hunk 2]
```diff
@@ -309,6 +309,77 @@ struct THCNumerics<float> {
...
+  static inline __host__ __device__  at::BFloat16 mul  (at::BFloat16 a, at::BFloat16 b) { return a * b; }
+  static inline __host__ __device__  at::BFloat16 sub  (at::BFloat16 a, at::BFloat16 b) { return a - b; }
+  static inline __host__ __device__  at::BFloat16 pow  (at::BFloat16 a, at::BFloat16 b) { return powf(a, b); }
+  static inline __host__ __device__  at::BFloat16 atan2(at::BFloat16 a, at::BFloat16 b) { return atan2f(a, b); }
+
+  static inline __host__ __device__ bool isnan(at::BFloat16 a) {
+    #ifdef _MSC_VER
+      // Windows requires this explicit conversion. The reason is unclear
+      // related issue with clang: https://reviews.llvm.org/D37906
+      return ::isnan((float) a);
```
Comment (ezyang) [inline aten/src/THC/THCNumerics.cuh]: nit: (you don't have to do anything about this, just fyi for later) generally people prefer `static_cast<float>` in C++ as it prevents the C style cast from, e.g., being treated like a `reinterpret_cast`

**Review by zasdfgbnm — COMMENTED** [review]
@izdeby I see your fix merge conflict commit https://github.com/pytorch/pytorch/pull/27259/commits/b34c8ddce4a0734c98faa22bb581f1e08b89668b
you simply delete the `BinaryOpsKernel.cu`. Is this correct? This file has been split due to its long compilation time. I guess you might need to modify the split files separately. See: https://github.com/pytorch/pytorch/pull/29428

**Comment by izdeby** [comment]
@zasdfgbnm, thanks! good catch

**Pull request closed without merging**